pkg/agent/multicast/multicast_route.go

```diff
@@ -0,0 +1,350 @@
+	g := group.To4()
+	mc.mfccMcastgrp = [4]byte{g[0], g[1], g[2], g[3]}
+	ttls := [32]byte{}
+	for _, v := range oifsVif {
+		ttls[v] = 1
+	}
+	mc.mfccTtls = ttls
+	mc.mfccParent = iif
+	return setsockoptMfcctl(c.sockFD, MRT_ADD_MFC, mc)
+}
+
+func (c *MRouteClient) delMrouteEntry(src net.IP, group net.IP, iif string) (err error) {
+	mc := &mfcctl{}
+	origin := src.To4()
+	mc.mfccOrigin = [4]byte{origin[0], origin[1], origin[2], origin[3]}
+	g := group.To4()
+	mc.mfccMcastgrp = [4]byte{g[0], g[1], g[2], g[3]}
+	vif, err := c.vifAllocator.setVifIfNotExist(iif)
+	if err != nil {
+		return err
+	}
+	mc.mfccParent = uint16(vif)
+	return setsockoptMfcctl(c.sockFD, MRT_DEL_MFC, mc)
+}
+
+func (c *MRouteClient) flushMRoute() (err error) {
+	klog.Infof("Clearing multicast routing table entry")
+	return setsockoptVifctl(c.sockFD, MRT_FLUSH, &vifctl{})
+}
+
+func (c *MRouteClient) DeleteMrouteEntryByGroup(group net.IP) (err error) {
+	klog.V(2).Infof("Deleting group %s", group.String())
+
+	mEntries, _ := c.inboundRouteCache.ByIndex(GroupNameIndexName, group.String())
+	for _, route := range mEntries {
+		entry := route.(*inboundMulticastRouteEntry)
+		err := c.delMrouteEntry(net.ParseIP(entry.src), net.ParseIP(entry.group), entry.iif)
+		if err != nil {
+			return err
+		}
+		c.inboundRouteCache.Delete(route)
+	}
+	return nil
+}
+
+func (c *MRouteClient) addOutBoundMrouteEntry(src net.IP, group net.IP, oifs []string) (err error) {
+	klog.InfoS("Adding outbound multicast route", "src", src, "group", group, "oifs", oifs)
+	oifsVif := make([]int, 0)
+	for _, oif := range oifs {
+		oifvif, err := c.vifAllocator.setVifIfNotExist(oif)
+		if err != nil {
+			return err
+		}
+		oifsVif = append(oifsVif, int(oifvif))
+	}
+
+	inboundVif, err := c.vifAllocator.setVifIfNotExist(c.gatewayInterface)
+	if err != nil {
+		return err
+	}
+	err = c.addMrouteEntry(src, group, uint16(inboundVif), oifsVif)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *MRouteClient) addInboundMrouteEntry(src net.IP, group net.IP, inboundVif uint16) (err error) {
+	klog.InfoS("Adding inbound multicast route", "src", src, "group", group, "inboundVif", inboundVif)
+
+	oif, err := c.vifAllocator.setVifIfNotExist(c.gatewayInterface)
+	if err != nil {
+		return err
+	}
+
+	err = c.addMrouteEntry(src, group, inboundVif, []int{int(oif)})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type inboundMulticastRouteEntry struct {
+	group string
+	src   string
+	iif   string
+}
+
+func getMulticastEntryKey(obj interface{}) (string, error) {
+	entry := obj.(*inboundMulticastRouteEntry)
+	return entry.group + "/" + entry.src + "/" + entry.iif, nil
+}
+
+func groupIndexFunc(obj interface{}) ([]string, error) {
+	entry, ok := obj.(*inboundMulticastRouteEntry)
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{entry.group}, nil
+}
+
+func (c *MRouteClient) getExternalMulticastAddrs() (map[string][]net.IP, error) {
```

Comment:
Could you add some comments to explain what does this func do?  In general I think you should add more comments to code for readers/maintenance.